### ISSUE TITLE
Keybindings

### ISSUE THREAD

**Issue by tecosaur** [James-Yu/LaTeX-Workshop#809]
# Background

Keybindings are great, we all love them, and LaTeX workshop currently doesn't have any. While this may not be as significant a change as other additions, after discovering [this](https://ftp.gnu.org/gnu/auctex/11.88-extra/tex-ref.pdf) since it seemed reasonably easy to implement (for the most part) I thought why not add it to the LaTeX workshop!
I found that `Ctrl`+`L` seems to be unused, why not use that instead of `C-c` from the AUCTEX reference card.

# Implementation Progress

This is a modified version of the AUCTEX reference card, some things removed, some things added.

### Issue encountered

- VS Code only take 2 part key bindings, not 3.

### _Command Insertion_

| Shortcut      | Does                     | Implemented |
| ------------- | ------------------------ | :---------: |
| `C-l C-s`     | Insert Section           |  Need Help  |
| `C-l C-e`     | Insert LaTeX Environment |  Need Help  |
| `C-l C-Enter` | Insert `\item`           |      ✓      |
| `M-Enter` | Insert `\item`           |      ✗      |
| `C-l Escape`  | Close LaTeX Environment  |  Need Help  |
| `C-l C-.` | Insert `\@.` | ✗ |
| `C-l C-space` | Insert <code>\\&nbsp;</code> | ✗ |

### _Font Selection_

| Shortcut Opt#1 | Shortcut Opt#2 | Does            | Implemented |
| -------------- | -------------- | --------------- | :---------: |
| `C-b`          | `C-l C-b`      | Bold text       |      ✓      |
| `C-i`          | `C-l C-i`      | Italic Text     |      ✓      |
| `C-u`          | `C-l C-u`      | Underline       |      ✓      |
| `C-r`          | `C-l C-r`      | Roman Text      |      ✓      |
| `C-e`          | `C-l C-e`      | Emphasised Text |      ✓      |
| `C-t`          | `C-l C-t`      | Typewriter Text |      ✓      |
| `C-s`          | `C-l C-s`      | Slanted Text    |      ✓      |
|                | `C-l C-c`      | Small Caps Text |      ✓      |
|                | `C-l C-l`      | Lowercase       |      ✓      |
|                | `C-l C-u`      | Uppercase       |      ✓      |
|                |                | Delete Font     |  Need help  |
|                |                | Replace Font    |  Need help  |

I'm currently using Opt#1 for BIU and emphasis when text is selected, then Opt#2 for everything else

| Shortcut   | Does            | Implemented |
| ---- | ---- | :---: |
| `C-l C-6` | `\textsuperscript` |     ✓      |
| `C-l C-dash` | `\textsubscript` |     ✓      |

Why the `6` and `dash`? Because they are the keys for `^` and `_`.

### _Source Selection_

Most of this is just in VS Code.

## Math Environment

AUCTEX uses the same prefix key for all it's math commands - `` ` `` would it be worth / can we copying this. Upon reflection it does seem a good key to use.

Actually upon coming back to this, I personally believe that I'd find the utility added by being able to use this as a math prefix key greater that the benefit of having `` ` `` automatically do the quote thing ( `` ` `` → `` `' `` ), particular since for double quotes that isn't actually that helpful.

*Edit:* Since this is an important deadkey for some keyboards sticking with `@` or having it be a setting would be a better idea.

### _Font Selection_

| Shortcut    | Does                  | Implemented |
| ----------- | --------------------- | :---------: |
| `C-m C-S-b` | Blackboard Font (𝔹)   |      ✓      |
| `C-m C-c`   | Calligraphic Font (𝓒) |      ✓      |
| `C-m C-f`   | Frak Font (𝔉)         |      ✓      |
| `C-m C-b`   | Bold                  |      ✓      |
| `C-m C-i`   | Italic                |      ✓      |
| `C-m C-r`   | Roman                 |      ✓      |
| `C-m C-t`   | Typewriter            |      ✓      |
| `C-m C-s`   | Sans Serif            |      ✓      |

### _Greek Letters_

We've already got this, but it might be worth changing `@f` → `\phi` instead of `\varphi`, for the variant letters perhaps something like `@vf` would be better? Also, if we can/do end up using `` ` `` as a math-shortcut-prefix, it may be worth using using that instead of @.

*Edit:* Done ✓

### _Symbols_

NB: I'm currently looking for a way to differentiate math mode from text mode, this will affect these shortcuts.
Update: Looks like this isn't possible.

| Shortcut                   | Does          | Implemented | Snippet |
| -------------------------- | ------------- | :---------: | ------- |
| `C-?? C-f`                 | `\rightarrow` |      ✓      | `@->`   |
| `C-?? C-b`                 | `\leftarrow`  |      ✓      | `@<-`   |
| `C-?? C-p`                 | `\uparrow`    |      ✓      |
| `C-?? C-n`                 | `\downarrow`  |      ✓      |
| `C-?? <`                   | `\leq`        |      ✓      | `@<`    |
| `C-?? >`                   | `\geq`        |      ✓      | `@>`    |
| `C-?? ~`                   | `\tilde`      |      ✓      | `@~`    |
| `C-?? ^`                   | `\hat`        |      ✓      | `@^`    |
| `C-?? N`                   | `\nalba`      |      ✓      | `@N`    |
| `C-?? I`                   | `\infty`      |      ✓      | `@I`    |
| `C-?? A`                   | `\forall`     |      ✓      | `@A`    |
| `C-?? E`                   | `\exists`     |      ✓      | `@E`    |
| `C-?? /`                   | `\not`        |      ✓      | `@/`    |
| `C-?? i`                   | `\in`         |      ✓      |
| `C-?? *`                   | `\times`      |      ✓      | `@*`    |
| `C-?? .`                   | `\cdot`       |      ✓      | `@.`    |
| `C-?? :`                   | `\colon`      |      ✓      | `@:`    |
| `C-?? {`                   | `\subset`     |      ✓      |
| `C-?? }`                   | `\supset`     |      ✓      |
| `C-?? [`                   | `\subseteq`   |      ✓      |
| `C-?? ]`                   | `\supseteq`   |      ✓      |
| `C-?? 0`                   | `\emptyset`   |      ✓      |
| `C-?? \` | `\setminus` | ✓ |
| `C-?? +`                   | `\cup`        |      ✓      | `@+`    |
| `C-?? -`                   | `\cap`        |      ✓      | `@-`    |
| `C-?? (`                   | `\langle`     |      ✓      |
| `C-?? )`                   | `\rangle`     |      ✓      |
| `C-?? C-e`                 | `\exp`        |      ✓      |
| `C-?? C-s`                 | `\sin`        |      ✓      |
| `C-?? C-c`                 | `\cos`        |      ✓      |
| `C-?? C-t`                 | `\tan`        |      ✓      |
| `C-?? C-^`                 | `\sup`        |      ✓      |
| `C-?? C-_`                 | `\inf`        |      ✓      |
| `C-?? C-d`                 | `\det`        |      ✓      |
| `C-?? C-l`                 | `\lim`        |      ✓      |
| <code>C-?? &#124;</code>   | `\vee`        |      ✓      | `@|`    |
| `C-?? &`                   | `\wedge`      |      ✓      | `@&`    |

### More Maths Ideas
| Shortcut                   | Does          | Implemented |
| -------------------------- | ------------- | :---------: |
| `C-m C-/` | `\frac{}{}` | ✗ |
| `C-m C-6` | `^{}` | ✗ |
| `C-m C-dash` | `_{}` | ✗ |
| `C-m C-w` | Wrap with  `\(…\)` | ✗ |

NB: looking for feedback on whether these are a good/bad/mediocre idea.

## Other stuff

- AUCTEX has other shortcuts, many of them cool. But they're not as simple to implement, and I've already spent long enough writing this our.
- I feel like this doesn't quite go in 'issues' - is there a better place for me to put it?
- @James-Yu I'm doing this by editing my local copy of this extension, as such it would be great if you could ping me before you release the next update so that I know to save what I've done so it doesn't get overwritten :)

**Comment by tecosaur** [comment]
@James-Yu how can I disable the ` `' `` thing for testing?

**Comment by James-Yu** [comment]
You may find luck in `LaTeX-Workshop/syntax/syntax.json` @tecosaur

**Comment by tecosaur** [comment]
Would you know of any particular keywordsi should search for? ` pops up a fair bit.

**Comment by James-Yu** [comment]
I'm afraid I don't get what you mean. If you are referring to the pairing of ` and ', the file is it:

```
	"surroundingPairs": [
		["{", "}"],
		["[", "]"],
		["(", ")"],
		["\"", "\""],
		["'", "'"],
		["`", "'"],
		["$", "$"]
	],
```

**Comment by tecosaur** [comment]
Oops I looked in Latex.plist
Thanks!

**Comment by tecosaur** [comment]
Oops I looked in Latex.plist
Thanks!

**Comment by jlelong** [comment]
Actually, `` ` `` is a dead key on some keyboard layouts (the French one for example), which makes barely unusable as a _map leader_. This is the reason, I have decided to use `@` for greek letters. I had exactly the same problem with _Auctex_ and _vim-latex_ and I had to change the map leader. Might be worth thinking of it before starting.

**Comment by tecosaur** [comment]
@jlelong I'd completely forgotten about that! Thanks for reminding me. With that in mind do you know if it would be possible to have the prefix key be changeable by a setting?

**Comment by jlelong** [comment]
I do not think you can have part of the prefix configurable in a snippet. It could be done by using a similar approach as for `\` now and as it is coded directly in typescript, the map leader can of course be configurable.

**Comment by tecosaur** [comment]
This would not be for a snippet but a keybinding - does that change anything?

**Comment by jlelong** [comment]
Hum, `@a` to get `\alpha` is a snippet as I understand well (see [here](https://code.visualstudio.com/docs/getstarted/keybindings)) you cannot define a keybinding not starting in a key modifier.

Anyway, it seems you refer to configuration variable using `${config:latex-workshop.mapLeader}` for instance.

**Comment by tecosaur** [comment]
Ahh I'm adding this as all keybindings, but having those things as snippets also makes (more) sense. Hadn't taken a moment to stop and think. Proceeding to do the math symbols as snippets.

(From a 2-minute google search) Would it be possible to register snippet completion item providers on extension load or similar?

**Comment by jlelong** [comment]
> (From a 2-minute google search) Would it be possible to register snippet completion item providers on extension load or similar?

Yes, this is already done in the extension for showing snippets suggestions on pressing `\`.

**Comment by tecosaur** [comment]
So that's what you've done. That's very promising :) I'll have a look and see what I can figure out from looking at the code. In anticipation of future-me having issues, which file might I find this in?

Also, I can't seem to find any way of having 'or' statements in the `when` field for keybindings - do I have to just have near-duplicates?

**Comment by jlelong** [comment]
Right, no `or` in `when`.
Look at `src/completion.ts`.

**Comment by jlelong** [comment]
I suggest to first deal with the snippets for math symbols and greek letters and see later if we can assign them keybindings (coming from the Vim world, I personally do not like keybindings that much).

_Vim-LaTeX_ and _Auctex_ use the same snippets for symbols and maths, see [here](http://vim-latex.sourceforge.net/documentation/latex-suite/auc-tex-mappings.html). I will integrate them into the extension and see if you can configure the map leader.

**Comment by tecosaur** [comment]
I've already added quite a few snippets - see the updated original post. If you want I'll update my fork and give the link to the new latex.json

**Comment by jlelong** [comment]
I have cherry-picked your commit and added some missing snippets from [here](http://vim-latex.sourceforge.net/documentation/latex-suite/auc-tex-mappings.html)

**Comment by jlelong** [comment]
I have just added a documentation for all the snippets available so far.
Variants of greek letters are available as `@v`+ `letter`.

**Comment by tecosaur** [comment]
For text formatting [here's](https://pastebin.com/8L3vHKVc) what I've got going at the moment. I'll probably make a pull request in the near future.

**Comment by jlelong** [comment]
It looks great. Waiting for your PR. I am wondering whether all these keybindings could be put in a dedicated `.json` file in order to keep  the size of `package.json` reasonable (it is already more than 700 lines).

**Comment by tecosaur** [comment]
Done :) Also would you happen to know if there's a similar way to reference the defined latex commands such as `\frac{}{}` that have placeholders without defining separate snippets for them?

**Comment by jlelong** [comment]
@tecosaur I have already added some stuff about snippets in [snippets-doc.md](https://github.com/James-Yu/LaTeX-Workshop/blob/master/snippets/snippets-doc.md) and there is a link to it in [readme.md](https://github.com/James-Yu/LaTeX-Workshop/blob/master/readme.md). It may be worth it completing it.

**Comment by tecosaur** [comment]
Sure:)

**Comment by tecosaur** [comment]
@jlelong Don't close this quite yet! There are a few more things I'm planning/considering...

**Comment by Astrantia** [comment]
@tecosaur How do I make use of the pull request you made? Clicking the key bindings don't work. I compiled it manually.

**Comment by tecosaur** [comment]
@Astrantia 

![shortcut demo](https://user-images.githubusercontent.com/20903656/46277196-05a07180-c595-11e8-80b6-dba651d83fdf.gif)

**Comment by tecosaur** [comment]
Aaaaah, I see your issue.
This may have something to do with the fact that I just copy and pasted from my local `keybindings.json`. It looks like some modification is required to make them work the same in `package.json`. I'll see what can be done. @James-Yu @jlelong you should probably be made aware of this.

**Comment by jlelong** [comment]
Bad news! This means we cannot easily provide keybindings.
As it seems to be a limitation of vscode, maybe we could ship a `keybindings.json` file with the extension containing the required definitions and explain the user thay he should include it in its own keybindgngs file. This is far from being optimal but I can't see any other simple solution.

**Comment by Astrantia** [comment]
@tecosaur Can you share your keybindings.json? 😁

**Comment by Astrantia** [comment]
@tecosaur @jlelong @James-Yu Actually the guy who opened that issue is the author of markdown all in one extension and I remembered markdown all in one extension supports selecting a text and making it bold. So I checked that extension out to find how he does it, it's here:
https://github.com/neilsustc/vscode-markdown/blob/master/package.json#L59

**Comment by jlelong** [comment]
Now, the keybindings are working. A new mechanism is implemented 8a40cb3 to define keybindings for the snippets in [snippets/latex.json](https://github.com/James-Yu/LaTeX-Workshop/blob/master/snippets/latex.json)

**Comment by tecosaur** [comment]
Fantastic! While that was going on I was looking at other possibilities I came across what seems like a superior implementation of this idea to mine, and the author seems interested in adding the functionality to LaTeX workshop.

[viktorstrate/vscode-latex-utilities#5](https://github.com/viktorstrate/vscode-latex-utilities/issues/5)
